README.md
# Free/Busy Calendar - https://www.porcheron.uk/fbc

## Usage
An example usage can be found in `example.php`. This file connects to a CalDAV server, extracts two weeks of dates 
(excluding weekends) and generates an HTML table. A sample walkthough of the code is below:

If you are loading an iCal file, use the `Calendar` class:

    $cal = (new MPorcheron\FreeBusyCal\Calendar())->setFile('calendar.ics');

Alternatively:

 * if the file is accessed over the internet, use the `setUrl(url)` function instead of `setFile(file)`. 
 * if you have the calendar source inside a string, use the function `setiCal(source)`

Howevever, if your calendar is retrieved from a CalDAV server and not an ICS file, use the `CalDAVCalendar` class:

    $iCloud = (new MPorcheron\FreeBusyCal\CalDAVCalendar())
        ->setUsername('my.apple.id@me.com')
        ->setPassword('application-specific-password')
        ->setPrincipalUrl('https://caldav.icloud.com/123456789876543/principal/');

Create the `Generator` object and add one or more calendars:

	$fbc = new \MPorcheron\FreeBusyCal\Generator($cal, $iCloud);

Set the date range to extract, e.g. start from this Monday, and run for 14 days (i.e. two weeks), but exclude
weekends:

    $fbc->setDateRange(new \DateTime('Monday this week'), 14, false);


Only generate a calendar between 9am (inclusive) and 5pm (exclusive):

    $fbc->setTimeRange(9, 17);


Fetch the calendars and process them:

    $fbc->fetchAndParse();


Print out the calendar as a table, default date and time formats, the labels `Free` and `Busy` for
slots, and show times as ranges (i.e. start – end):

    $contents = $fbc->generate(function (Fbc\FreeBusyCalendar &$cal) {
        $output = '<table class="cal">';

        // Output table headers with days
        $output .= '<tr><th></th>';
        $days = [ 'S', 'M', 'T', 'W', 'T', 'F', 'S' ];
        foreach ($cal->getCalendarDates(Fbc\FreeBusyCalendar::DATE_FORMAT) as $label => &$dt) {
            $output .= '<th class="day">'. $days[$dt->format('N')] .'</th>';
        }
        $output .= '</tr>';

        // Output table headers with dates
        $output .= '<tr><th></th>';
        foreach ($cal->getCalendarDates(Fbc\FreeBusyCalendar::DATE_FORMAT) as $label => &$dt) {
            $output .= '<th class="date">'. $label .'</th>';
        }
        $output .= '</tr>';

        // Iterate through each time and $output .= the availability
        $times = $cal->getCalendarTimes(Fbc\FreeBusyCalendar::TIME_FORMAT);
        foreach ($times as $hour => $temp) {
            foreach ($temp as $minute => $labels) {
                $output .= '<tr><td class="time">'. $labels[0];
                if ($showRange) {
                    $output .= '&nbsp;&ndash;&nbsp;' . $labels[1];
                }
                $output .= '</td>';

                foreach ($cal->getCalendarDates(Fbc\FreeBusyCalendar::DATE_FORMAT) as $dt) {
                    if ($cal->isFree($dt->format('Y-m-d'), $hour, $minute)) {
                        $output .= '<td class="avail free">Free</td>';
                    } else {
                        $output .= '<td class="avail busy">Busy</td>';
                    }
                }
            }
            $output .= '</td>';
        }
        $output .= '</table>';

        return $output;
    });


Alternatively test if a specific time/date (i.e. 5pm on 4th May 2016) is available:

     $cal = $fbc->getFreeBusyCalendar();
     $free = $cal->isFree('2016-05-04', 17, 0);


## Testing
This has been tested with the ICS file from Office 365 and iCloud CalDAV. 

## Questions/Issues?
Please submit a [GitHub issue](https://github.com/mporcheron/free-busy-cal). Documentation auto-generated from the 
code can be found on [GitHub pages](http://mporcheron.github.io/free-busy-cal/) or the `docs` directory.


CHANGELOG
v0.4.1
------
- Ready the codebase for Packagist
- Add contributing information to the repository

v0.4
----
- Rewrite CalDAV functionality with Sabre\DAV client (now fully using composer)
- Rename 'UserCalendar' back to 'Calendar'
- Move CalDAV functionality to subclass of 'Calendar', 'CalDAVCalendar'
- Pass configuration to the fetch() function of 'Calendar' class
- Fix: Merged calendars would overwrite each other
- Fix: Generating output with multiple calendars
- Fix: Added missing changelog

v0.3.1
------
- Fix the example code

v0.3
----
- Calendar source data is now properly encapsulated in UserCalendar 
- Calendar printing is now easier to customise.
- Improved commenting and documentation
- Support minutes in time ranges
- Renamed the package
- Update Composer and add sabre\dav for future use
- Fix: Values for time ranges weren't treated as documented, fixed code
- Fix: merging calendars was not always correct

v0.2
----
- Update dependencies
- Add composer support properly
- Begin modularisation of the code
- Added example walkthrough
- Clearer commenting and documentation

v0.1
----
- Initial release


src/Generator.php
<?php

/**
 * This file is the main FreeBusyCal generator used to retrieve and generate a free/busy calendar. See the class
 * description for example usage.
 *
 * @package MPorcheron\FreeBusyCal
 */

namespace MPorcheron\FreeBusyCal;

/**
 * FreeBusyCal generator - this is the main class used to generate a free busy calendar.
 *
 * If you are loading an iCal file, use the `Calendar` class:
 *
 *      $cal = (new MPorcheron\FreeBusyCal\Calendar())->setFile('calendar.ics');
 *
 * Alternatively:
 * <ul>
 * <li>if the file is accessed over the internet, use the `setUrl(url)` function instead of `setFile(file)`.</li>
 * <li>if you have the calendar source inside a string, use the function `setiCal(source)`</li>
 * <li>If your calendar is retrieved from a CalDAV server, use the `CalDAVCalendar` class:</li>
 * </ul>
 *
 *      $iCloud = (new MPorcheron\FreeBusyCal\CalDAVCalendar())
 *          ->setUsername('my.apple.id@me.com')
 *          ->setPassword('application-specific-password')
 *          ->setPrincipalUrl('https://caldav.icloud.com/123456789876543/principal/');
 *
 * Create the `Generator` object and add one or more calendars:
 *
 *     $fbc = new \MPorcheron\FreeBusyCal\Generator($cal, $iCloud);
 *
 * Set the date range to extract, e.g. start from this Monday, and run for 14 days (i.e. two weeks), but exclude
 * weekends:
 *
 *      $fbc->setDateRange(new \DateTime('Monday this week'), 14, false);
 *
 *
 * Only generate a calendar between 9am (inclusive) and 5pm (exclusive), and show a slot every 30 minutes:
 *
 *      $fbc->setTimeRange(9, 17, 30);
 *
 *
 * Fetch the calendars and process them:
 *
 *      $fbc->fetchAndParse();
 *
 *
 * Print out the calendar table, with the class `cal`, default date and time formats, the labels `Free` and `Busy` for
 *  slots, and show times as ranges (i.e. start – end) as opposed to just start time:
 *
 *     $contents = $fbc->generate(function (Fbc\FreeBusyCalendar &$cal) {
 *         $output = '<table class="cal">';
 *
 *         // Output table headers with days
 *         $output .= '<tr><th></th>';
 *         $days = [ 'S', 'M', 'T', 'W', 'T', 'F', 'S' ];
 *         foreach ($cal->getCalendarDates(Fbc\FreeBusyCalendar::DATE_FORMAT) as $label => &$dt) {
 *             $output .= '<th class="day">'. $days[$dt->format('N')] .'</th>';
 *         }
 *         $output .= '</tr>';
 *
 *         // Output table headers with dates
 *         $output .= '<tr><th></th>';
 *         foreach ($cal->getCalendarDates(Fbc\FreeBusyCalendar::DATE_FORMAT) as $label => &$dt) {
 *             $output .= '<th class="date">'. $label .'</th>';
 *         }
 *         $output .= '</tr>';
 *
 *         // Iterate through each time and $output .= the availability
 *         $times = $cal->getCalendarTimes(Fbc\FreeBusyCalendar::TIME_FORMAT);
 *         foreach ($times as $hour => $temp) {
 *             foreach ($temp as $minute => $labels) {
 *                 $output .= '<tr><td class="time">'. $labels[0];
 *                 if ($showRange) {
 *                     $output .= '&nbsp;&ndash;&nbsp;' . $labels[1];
 *                 }
 *                 $output .= '</td>';
 *
 *                 foreach ($cal->getCalendarDates(Fbc\FreeBusyCalendar::DATE_FORMAT) as $dt) {
 *                     if ($cal->isFree($dt->format('Y-m-d'), $hour, $minute)) {
 *                         $output .= '<td class="avail free">Free</td>';
 *                     } else {
 *                         $output .= '<td class="avail busy">Busy</td>';
 *                     }
 *                 }
 *             }
 *             $output .= '</td>';
 *         }
 *         $output .= '</table>';
 *
 *         return $output;
 *     });
 *
 * Alternatively test if a specific time/date (i.e. 5pm on 4th May 2016) is available:
 *
 *      $cal = $fbc->getFreeBusyCalendar();
 *      $free = $cal->isFree('2016-05-04', 17, 0);
 *
 *
 * @author    Martin Porcheron <martin-fbc@porcheron.uk>
 * @copyright (c) Martin Porcheron 2017.
 * @license   MIT Licence
 */

class Generator
{
    
    /**
     * @var \MPorcheron\FreeBusyCal\Calendar[] Array of calendars to scrape for data.
     */
    private $calendars;

    /**
     * @var mixed[] Configuration data.
     */
    private $config;

    /**
     * @var \MPorcheron\FreeBusyCal\FreeBusyCalendar Cached calendar data.
     */
    private $freeBusyCalendar = null;

    /**
     * Construct the controller and populate it with the configuration values. Constructing this class sets the time
     * limit for script execution to indefinite and the default timezone because of a PHP oddity.
     *
     * Output buffering is started also.
     *
     * @param \MPorcheron\FreeBusyCal\Calendar $cal
     *  (One or more) calendars to extract data from.
     */
    public function __construct(&$cal)
    {
        \set_time_limit(0);
        \date_default_timezone_set('Europe/London');

        $this->calendars = \func_get_args();

        $this->config  = [
            'numDays' => 14,
            'startDate' => new \DateTime('Monday this week'),
            'endDate' => new \DateTime('Sunday this week'),
            'startHour' => 9,
            'endHour' => 17,
            'interval' => 60,
            'includeWeekends' => false];
    }

    /**
     * Get the congiruation data.
     *
     * @return mixed[]
     *  Configuration data.
     */
    public function getConfig()
    {
        return $this->config;
    }

    /**
     * Add a calendar to extract data from.
     *
     * @param \MPorcheron\FreeBusyCal\Calendar $cal
     *  Calendar to also extact data from.
     */
    public function addCalendar(Calendar &$cal)
    {
        $this->calendars[] = $cal;
    }

    /**
     * Set the date range to generate data for.
     *
     * @see    http://php.net/manual/en/datetime.construct.php
     * @param  \DateTime $startDate
     *  When to start the calendar from.
     * @param  int       $numDays
     *  Number of days from the start date to generate calendar for.
     * @param  boolean   $includeWeekends
     *  Set to false to ignore weekends. Note weekennds still count in the number of days (i.e. 7 days, and
     *  `$includeWeekends` starting on a Monday, will show Mon - Fri.)
     * @return \MPorcheron\FreeBusyCal\Calendar
     *  `$this`.
     */
    public function setDateRange(\DateTime &$startDate, $numDays = 7, $includeWeekends = false)
    {
        $this->config['startDate'] = $startDate;
        $this->config['endDate'] = clone $startDate;

        $this->config['numDays'] = \filter_var(
            $numDays,
            FILTER_SANITIZE_NUMBER_INT,
            ['options' => ['default' => 7, 'min_range' => 1]]
        );
        $this->config['endDate']->add(new \DateInterval('P' . $this->config['numDays'] . 'D'));

        $this->config['includeWeekends'] = \filter_var($includeWeekends, FILTER_VALIDATE_BOOLEAN);

        return $this;
    }

    /**
     * Set the time range to generate data for.
     *
     * @see    http://php.net/manual/en/datetime.construct.php
     * @param  int $startHour
     *  First hour of the day to start output at (midnight = `0`). Minimum is `1`, maximum is `22`, default is `9`.
     * @param  int $endHour
     *  Last hour of the day print output for (midnight = 0). Minimum is `$startHour`, maximum is `23`, default is `17`.
     * @param  int $interval
     *  How many mniutes to break each slot in the calendar up by (60 = segment by hour). You should make this number
     *  one of the following to fit evenly into the hour: 1,2,3,4,5,6,10,12,15,20,30,60. Minumum is `1`, maximum is
     *  `60`, default is `60`.
     * @return \MPorcheron\FreeBusyCal\Generator
     *  `$this`.
     */
    public function setTimeRange($startHour, $endHour, $interval = 60)
    {
        $startHour = \filter_var(
            $startHour,
            FILTER_SANITIZE_NUMBER_INT,
            ['options' => ['default' => 9, 'min_range' => 1, 'max_range' => 22]]
        );
        $this->config['startHour'] = $startHour;

        $endHour = \filter_var(
            $endHour,
            FILTER_SANITIZE_NUMBER_INT,
            ['options' => ['default' => 17, 'min_range' => $startHour, 'max_range' => 23]]
        );
        $this->config['endHour'] = $endHour;

        $interval = \filter_var(
            $interval,
            FILTER_SANITIZE_NUMBER_INT,
            ['options' => ['default' => 60, 'min_range' => 1, 'max_range' => 60]]
        );
        $this->config['interval'] = $interval;

        return $this;
    }

    /**
     * Fetch and parse the data needed to generate the availability calendar.
     *
     * @param  boolean $refetch
     *  Refetch iCal data if it has already been fetched once.
     * @return \MPorcheron\FreeBusyCal\Generator
     *  `$this`.
     */
    public function fetchAndParse($refetch = false)
    {
        // Clear object data caches
        $this->freeBusyCalendar = null;

        // Fetch data from the CalDAV server
        $freeBusyCalendar = null;
        foreach ($this->calendars as $cal) {
            if (is_null($freeBusyCalendar)) {
                $freeBusyCalendar = $cal->fetch($this->config, $refetch)->parse($this->config);
            } else {
                $freeBusyCalendar->merge($cal->fetch($this->config, $refetch)->parse($this->config));
            }
        }

        $this->freeBusyCalendar = $freeBusyCalendar->setConfig($this->config);

        return $this;
    }

    /**
     * Generates the calendar and returs output.
     *
     * @param  function $func
     *  Function that takes a single paramter of a \MPorcheron\FreeBusyCal\FreeBusyCalendar` and returns out the output
     *  as a string.
     * @return \MPorcheron\FreeBusyCal\Generator
     *  Output from the print function.
     * @throws \BadFunctionCallException
     *  If the calendar hasn't been fetched yet.
     */
    public function generate($func)
    {
        if (\is_null($this->freeBusyCalendar)) {
            throw new \BadFunctionCallException('Must call Generator::fetchAndParse() before querying availability');
        }

        return $func($this->freeBusyCalendar);
    }

    /**
     * Retrieve the calendar of availability.
     *
     * @return \MPorcheron\FreeBusyCal\FreeBusyCalendar
     *  Calendar availability.
     * @throws \BadFunctionCallException
     *  If the calendar hasn't been fetched yet.
     */
    public function getFreeBusyCalendar()
    {
        if (\is_null($this->freeBusyCalendar)) {
            throw new \BadFunctionCallException('Must call Generator::fetchAndParse() before querying availability');
        }

        return $this->freeBusyCalendar;
    }
}
